server/devpi_server/model.py

```diff
@@ -499,6 +525,8 @@ def get_link_from_entrypath(self, entrypath):
         return links[0] if links else None
 
     def store_toxresult(self, link, toxresultdata):
+        if self.customizer.readonly:
+            raise ReadonlyIndex("index is marked read only")
```

Comment:
I see no tests for ``ReadonlyIndex`` exceptions

Reply:
These are not reached by regular code, because the permissions for UnknownCustomizer shortcut it. These exceptions are thus an additional security net. Not sure whether we should explicitly test for them.

Reply:
The test which would reach these if the permissions wouldn't work is ``test_permissions_for_unknown_index``